File: patriot_center_frontend/src/services/options.js
import { apiGet } from '../config/api';

/**
 * Fetches valid filtering options from the API.
 *
 * @returns {Promise<Object>} A promise that resolves to an object containing valid filtering options.
 */
export async function fetchOptions() {
  return apiGet('/dynamic_filtering');
}

/**
 * Fetches valid filtering options from the API based on the provided parameters.
 *
 * @param {string|number|null} year - The year of the data to filter.
 * @param {string|number|null} week - The week of the data to filter.
 * @param {string|null} manager - The manager of the data to filter.
 * @param {string|null} player - The player of the data to filter.
 * @param {string|null} position - The position of the data to filter.
 * @returns {Promise<Object>} A promise that resolves to an object containing valid filtering options.
 */
export async function fetchDynamicFiltering(year = null, week = null, manager = null, player = null, position = null) {
  // Construct the query string
  const params = new URLSearchParams();

  // Append the year if provided
  if (year) params.append('yr', String(year));

  // Append the week if provided
  if (week) params.append('wk', String(week));

  // Append the manager if provided
  if (manager) params.append('mgr', String(manager));

  // Append the player if provided
  if (player) params.append('plyr', String(player));

  // Append the position if provided and not equal to 'ALL'
  if (position && position !== 'ALL') params.append('pos', String(position));

  // Construct the path
  const queryString = params.toString();
  const path = queryString ? `/dynamic_filtering?${queryString}` : '/dynamic_filtering';

  // Fetch the data
  return apiGet(path);
}

File: patriot_center_backend/routes/options.py
"""Options endpoints for Patriot Center."""

from flask import Blueprint, Response, jsonify, request

from patriot_center_backend.exporters.dynamic_filter_exporter import (
    get_dynamic_filter_options,
)
from patriot_center_backend.exporters.options_exporter import (
    get_options_list,
)
from patriot_center_backend.models import Player
from patriot_center_backend.utils.data_formatters import to_records

bp = Blueprint("options", __name__)


@bp.route("/options/list", methods=["GET"])
def get_options_list_route() -> tuple[Response, int]:
    """Endpoint to list all players and managers in the cache.

    Returns:
        Response in JSON format and status code.
    """
    data = get_options_list()

    if request.args.get("format") == "json":
        response = jsonify(data)

        # Cache for 1 hour
        response.headers["Cache-Control"] = "public, max-age=3600"
        return response, 200

    response = jsonify(to_records(data, key_name="name"))

    # Cache for 1 hour
    response.headers["Cache-Control"] = "public, max-age=3600"
    return response, 200


@bp.route("/dynamic_filtering", methods=["GET"])
def get_dynamic_filter_options_route() -> tuple[Response, int]:
    """Endpoint retreive filtered data for provided input combinations.

    Receive any combination of season, week, manager, player, and position
    and dynamically filter the possible valid inputs based on the inputs
    provided.

    Acceptable arguments:
    - yr: Season (year)
    - wk: Week number
    - mgr: Manager name
    - pos: Player position
    - plyr: Player ID

    Returns:
        Response in JSON format and status code.
    """
    acceptable_args = ["yr", "wk", "mgr", "pos", "plyr"]
    for arg in request.args:
        if arg not in acceptable_args:
            return jsonify({"error": f"Invalid argument: {arg}"}), 400

    yr = request.args.get("yr")
    wk = request.args.get("wk")
    mgr = request.args.get("mgr")
    pos = request.args.get("pos")
    plyr = request.args.get("plyr")

    player = Player(plyr) if plyr else None

    try:
        data = get_dynamic_filter_options(yr, wk, mgr, pos, player)
    except ValueError as e:
        return jsonify({"error": str(e)}), 400
    response = jsonify(data)

    # Cache for 1 hour
    response.headers["Cache-Control"] = "public, max-age=3600"
    return response, 200
